Orange/widgets/data/tests/test_owsave.py

```diff
@@ -0,0 +1,89 @@
+# Test methods with long descriptive names can omit docstrings
+# pylint: disable=missing-docstring
+from unittest.mock import patch
+
+from Orange.data import Table
+from Orange.data.io import TabReader, PickleReader, FileFormat
+from Orange.tests import named_file
+from Orange.widgets.tests.base import WidgetTest
+from Orange.widgets.utils.filedialogs import format_filter, fix_extension, open_filename_dialog_save
+from Orange.widgets.data.owsave import OWSave
+
+
+class AddedFormat(FileFormat):
+    EXTENSIONS = ('.234',)
+    DESCRIPTION = "Test if a dialog format works after reading OWSave"
+
+    def write_file(self):
+        pass
+
+
+class TestOWSave(WidgetTest):
+
+    def setUp(self):
+        self.widget = self.create_widget(OWSave)  # type: OWSave
+
+    def test_ordinary_save(self):
+        self.send_signal(self.widget.Inputs.data, Table("iris"))
+
+        for ext, writer in [('.tab', TabReader), ('.pickle', PickleReader)]:
+            with named_file("", suffix=ext) as filename:
+                def choose_file(a, b, c, d, e, fn=filename, w=writer):
+                    return fn, format_filter(w)
+                with patch("AnyQt.QtWidgets.QFileDialog.getSaveFileName", choose_file):
+                    self.widget.save_file_as()
+                self.assertEqual(len(Table(filename)), 150)
+
+    def test_filename_with_fix_extension(self):
+
+        def mock_fix_choice(ret):
+            f = lambda a, b, c, d: ret
+            for a, v in vars(fix_extension).items():
+                setattr(f, a, v)
+            return f
+
+        fix_change_ext = mock_fix_choice(fix_extension.CHANGE_EXT)
+        fix_change_format = mock_fix_choice(fix_extension.CHANGE_FORMAT)
+        fix_cancel = mock_fix_choice(fix_extension.CANCEL)
+
+        filename = "some_output.tab"
+
+        def pickle_ext_with_tab_filter(*args):
+            return filename.replace("tab", "pickle"), format_filter(TabReader)
+
+        def tab_ext_with_pickle_filter(*args):
+            return filename, format_filter(PickleReader)
+
+        counter = [0]
+
+        def tab_ext_and_change_filter(*args):
+            counter[0] += 1
```

Comment:
Again, use `nonlocal`.

I know this is not longer (it's even shorter), but it's not idiomatic.